src/hotspot/share/compiler/compilerOracle.cpp

```diff
@@ -410,15 +410,19 @@ bool CompilerOracle::should_break_at(const methodHandle& method) {
 }
 
 bool CompilerOracle::should_blackhole(const methodHandle& method) {
-  if (check_predicate(CompileCommand::Blackhole, method)) {
-    if (method->result_type() == T_VOID) {
-      return true;
-    } else {
-      warning("blackhole compile command only works for methods with void type: %s",
-              method->name_and_sig_as_C_string());
-    }
+  if (!has_option(method, CompileCommand::Blackhole)) {
+    return false;
   }
-  return false;
+  if (!UnlockDiagnosticVMOptions) {
```

Comment:
The check happens too late (during compilation when JIT encounters marked method).

IMO it should happen during command-line parsing at startup: `CompilerOracle::parse_from_line()` or `register_command()` look suitable.

Reply:
Right. I moved it to `register_command`, and `guarantee`-d in `CompilerOracle::should_blackhole` that we ever get there (that is, predicate matched) if diagnostic options were unlocked. Still passes tests in both release and fastdebug.